daemon/images/images.go

```diff
@@ -177,56 +175,60 @@ func (i *ImageService) Images(imageFilters filters.Args, all bool, withExtraAttr
 				if imageFilters.Contains("reference") { // skip images with no references if filtering by reference
 					continue
 				}
-				newImage.RepoDigests = []string{"<none>@<none>"}
-				newImage.RepoTags = []string{"<none>:<none>"}
+				summary.RepoDigests = []string{"<none>@<none>"}
+				summary.RepoTags = []string{"<none>:<none>"}
 			} else {
 				continue
 			}
-		} else if danglingOnly && len(newImage.RepoTags) > 0 {
+		} else if danglingOnly && len(summary.RepoTags) > 0 {
 			continue
 		}
 
 		if withExtraAttrs {
-			// lazily init variables
-			if imagesMap == nil {
+			// Lazily init variables
+			if summaryMap == nil && allContainers == nil {
+				summaryMap = make(map[*image.Image]*types.ImageSummary, len(selectedImages))
 				allContainers = i.containers.List()
-				allLayers = i.layerStore.Map()
-				imagesMap = make(map[*image.Image]*types.ImageSummary)
-				layerRefs = make(map[layer.ChainID]int)
 			}
 
 			// Get container count
-			newImage.Containers = 0
+			summary.Containers = 0
 			for _, c := range allContainers {
 				if c.ImageID == id {
-					newImage.Containers++
+					summary.Containers++
 				}
 			}
+			summaryMap[img] = summary
+		}
+
+		summaries = append(summaries, summary)
+	}
 
-			// count layer references
+	if withExtraAttrs {
+		allLayers := i.layerStore.Map()
+		layerRefs := make(map[layer.ChainID]int, len(allLayers))
+		for _, img := range allImages {
+			// Count layer references
 			rootFS := *img.RootFS
 			rootFS.DiffIDs = nil
 			for _, id := range img.RootFS.DiffIDs {
 				rootFS.Append(id)
 				chid := rootFS.ChainID()
+
 				layerRefs[chid]++
 				if _, ok := allLayers[chid]; !ok {
 					return nil, fmt.Errorf("layer %v was not found (corruption?)", chid)
 				}
 			}
-			imagesMap[img] = newImage
-		}
 
-		images = append(images, newImage)
-	}
+		}
 
-	if withExtraAttrs {
 		// Get Shared sizes
-		for img, newImage := range imagesMap {
+		for img, summary := range summaryMap {
 			rootFS := *img.RootFS
 			rootFS.DiffIDs = nil
 
-			newImage.SharedSize = 0
+			summary.SharedSize = 0
```

Comment:
Perhaps we should add a comment here to describe why we're setting it to `0` here, e.g.

```suggestion
			// Indicate that we collected shared size information (default is -1, or "not set")
			summary.SharedSize = 0
```